File: go/bob/bob.go
// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import s "strings"
import "unicode"

var TO_QUESTION = "Sure."
var TO_YELL = "Whoa, chill out!"
var TO_YELLED_QUESTION = "Calm down, I know what I'm doing!"
var CALL = "Fine. Be that way!"
var WHATEVER = "Whatever."

var QUESTION_MARK = "?"
var CUTSET = "\t\r\n "

func isQuestion(str string) bool {
	return s.HasSuffix(str, QUESTION_MARK)
}

func isYelled(str string) bool {
	return s.ToUpper(str) == str && hasCharacters(str)
}

func hasCharacters(str string) bool {
	for _, c := range str {
		if unicode.IsLetter(c) {
			return true
		}
	}
	return false
}

func Hey(remark string) string {
	trimmed_remark := s.TrimRight(s.TrimLeft(remark, CUTSET), CUTSET)
	if isQuestion(trimmed_remark) {
		if isYelled(trimmed_remark) {
			return TO_YELLED_QUESTION
		} else {
			return TO_QUESTION
		}
	}
	if isYelled(trimmed_remark) {
		return TO_YELL
	}

	if trimmed_remark == "" {
		return CALL
	} else {
		return WHATEVER
	}

	return ""
}

File: csharp/bob/Bob.cs
using System;
using System.Text.RegularExpressions;

public static class Bob
{
    
    static bool isYelled(string statement) =>
       (new Regex(@"[a-z]|^[^A-Za-z]*$|[a-z][A-Z]|[A-Z][a-z]", RegexOptions.Compiled))
        .Matches(statement).Count == 0;

    static bool isNothing(string statement) => 
        new Regex(@"^[^A-Za-z0-9]*$", RegexOptions.Compiled).Matches(statement).Count != 0;

    static bool isQuestion(string statement) =>
        statement.Length > 0 && statement[statement.Length - 1] == '?';

    public static string Response(string statement)
    {
        string s = statement.TrimStart().TrimEnd();
        bool yelled = isYelled(s);
        bool nothing = isNothing(s);
        bool question = isQuestion(s);
        if (question && yelled) return "Calm down, I know what I'm doing!";
        if (question && !yelled) return "Sure.";
        if (nothing) return "Fine. Be that way!";
        if (yelled) return "Whoa, chill out!";
        return "Whatever.";

    }
}

